[lldb] add stop-at-user-entry option to process launch

junior-jl opened the pull request:
## Description
This pull request adds a new `stop-at-user-entry` option to LLDB `process launch` command, allowing users to launch a process and pause execution at the entry point of the program (for C-based languages, `main` function).

## Motivation
This option provides a convenient way to begin debugging a program by launching it and breaking at the desired entry point.

## Changes Made
- Added `stop-at-user-entry` option to `Options.td` and the corresponding case in `CommandOptionsProcessLaunch.cpp` (short option is 'm')
- Implemented `GetUserEntryPointName` method in the Language plugins available at the moment.
- Declared the `CreateBreakpointAtUserEntry` method in the Target API.
- Create Shell test for the command `command-process-launch-user-entry.test`.


## Usage
`process launch --stop-at-user-entry` or `process launch -m` launches the process and pauses execution at the entry point of the program.

jimingham reviewed (changes requested):
Half of this command is just `process launch --stop-at-entry`.  If you wanted to can the "break at main" part, then add a --stop-at-main to process launch.  And if you want a "start" command, the lldb way to do that would be to make a regex command that either does `process launch --stop-at-entry` or the new `process launch --stop-at-main`.

We don't add real top level commands that are just aliases to extant behavior.

jimingham commented:
Even if this were enough behavior to warrant a new command, it would live under the `process launch` node, not as a top level command.  We try not to make lots of top level commands so as to leave plenty of namespace in the easy access commands for user customization.

junior-jl commented:
Thank you for the explanation here and in Discourse. I added the option `stop-at-main` to `process launch`.

jimingham reviewed (commented):
The reason that gdb has a "start" command is that not all languages and their runtimes use the same name for the the entry point function.  So `start` does "figure out what the start function in the main executable is and set a breakpoint there".  Or maybe it just knows the start points of all the languages and breaks symbolically on each one, I can't remember how gdb does this...
In our case, the languages we currently support all use "main" for the entry point function, but since we're going to the trouble of making an canned bit of functionality for this, it might be nice to add that abstraction.

jimingham commented:
I can't tell how to read the "Files Changed" part of the GitHub PR change.  It looks like the --stop-at-main is additive, but to be clear, the second of these two commits looks okay to me, but the first commit is still not something we want to do.

walter-erquinigo commented:
Btw, I'm building support for the mojo language, and very likely the `main` method will not have the symbol `main`, so building an abstraction around this, for every language, would be very convenient.

jimingham commented:
The other improvement I'd suggest here is that the breakpoints you set to implement the "stop at main" behavior need to be be "one-shot" breakpoints.  Otherwise, if you do:

(lldb) process launch --stop-at-main
... the process exits

and then:

(lldb) process launch

you're still going to stop at main, which is surprising.

junior-jl commented:
This might be a dumb question, but I committed the implementation as an option as the second commit thinking you would/could discard my first commit. Is that possible? What would be the best way to correct this?

walter-erquinigo commented:
That's file. You can have multiple commits in the same PR, because when it gets approved and merged, github will squash all your commits into a single one.

bulbazord commented:
Concretely, you can revert the first commit and push that so that your branch has your first commit, second commit, and a revert commit for the first one. As Walter said, it should squash down and resolve correctly.

bulbazord commented:
You also may want to change the title of your PR to reflect the actual change you'll be making as well. When you squash and merge, you'll have an opportunity to rewrite the commit message to more accurately reflect what your change does as well.

clayborg commented on lldb/source/Commands/CMakeLists.txt:
```diff
@@ -30,6 +30,7 @@ add_lldb_library(lldbCommands NO_PLUGIN_DEPENDENCIES
   CommandObjectSession.cpp
   CommandObjectSettings.cpp
   CommandObjectSource.cpp
+  CommandObjectStart.cpp
```
revert this change since we are just adding an option to "process launch" right?

clayborg commented on lldb/source/Commands/CommandObjectStart.cpp:
```diff
@@ -0,0 +1,102 @@
+//===-- CommandObjectStart.cpp -------------------------------------------===//
```
Revert this source file since we are just adding an option to "process launch" right?

clayborg commented on lldb/source/Commands/CommandObjectStart.h:
```diff
@@ -0,0 +1,32 @@
+//===-- CommandObjectStart.h ---------------------------------*- C++ -*-===//
```
Revert this source file since we are just adding an option to "process launch" right?

clayborg commented on lldb/source/Commands/CommandOptionsProcessLaunch.cpp:
```diff
@@ -38,7 +38,18 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
+    TargetSP target_sp =
+        execution_context ? execution_context->GetTargetSP() : TargetSP();
+    BreakpointSP bp_sp = target_sp->CreateBreakpoint(
+        nullptr, nullptr, "main", eFunctionNameTypeAuto, eLanguageTypeUnknown,
```
eFunctionNameTypeAuto is too broad and might not work for all languages. Most executable object files have an accessor to fetch the entry point address, which is the address of the main function. The "--stop-at-entry" is for stopping right away at the first instruction of the program, not the main function, so even though the API used below mentions the entry point, it is actually the main function.
```
ModuleSP exe_module_sp = target_sp->GetExecutableModule();
lldb_private::Address entry_addr;
if (exe_module_sp ) {
  ObjectFile *objfile = exe_module_sp ->GetObjectFile();
  if (objfile)
      entry_addr = objfile->GetEntryPointAddress();
}
if (entry_addr.IsValid()) {
  BreakpointSP bp_sp = target_sp->CreateBreakpoint(entry_addr, /*internal=*/false, /*hardware=*/false);
  if (!bp_sp)
    error.SetErrorString("entry breakpoint creation failed.");
} else {
  error.SetErrorString("executable doesn't have an entrypoint address.");
}
```

clayborg replied:
This actually only works on darwin, as the entry point address actually points to the main function. For linux, this points to "_start", so my suggested approach above doesn't actually work for all platforms.

junior-jl replied:
I resolved the other points, this is the only part missing. Should I use `eFunctionNameTypeBase`?

clayborg commented on lldb/source/Commands/CommandOptionsProcessLaunch.cpp:
```diff
@@ -38,7 +38,18 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
+        0, eLazyBoolNo, false, false);
+    if (!bp_sp)
+      error.SetErrorString("Breakpoint creation failed.\n");
+    launch_info.SetShell(HostInfo::GetDefaultShell());
```
remove this line. We shouldn't be modifying the shell in the entry point breakpoint case.

clayborg commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -36,6 +36,7 @@
 #include "Commands/CommandObjectSession.h"
 #include "Commands/CommandObjectSettings.h"
 #include "Commands/CommandObjectSource.h"
+#include "Commands/CommandObjectStart.h"
```
revert

clayborg commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -559,6 +560,7 @@ void CommandInterpreter::LoadCommandDictionary() {
   REGISTER_COMMAND_OBJECT("settings", CommandObjectMultiwordSettings);
   REGISTER_COMMAND_OBJECT("session", CommandObjectSession);
   REGISTER_COMMAND_OBJECT("source", CommandObjectMultiwordSource);
+  REGISTER_COMMAND_OBJECT("start", CommandObjectStart);
```
revert

junior-jl commented:
Thank you @bulbazord & @clayborg for the step-by-step, it was really helpful. I reverted the first commit, removed the shell modification you mentioned and the created breakpoint is now one-shot, @jimingham.

clayborg reviewed (commented):
See my inline comments. I am wondering if we want to get the platform involved here so they can do the right thing for any platform and it might involve renaming the option away from containing "main" since that might not be the user entry point into a program for all platforms.

jimingham commented on lldb/source/Commands/CommandOptionsProcessLaunch.cpp:
```diff
@@ -38,7 +38,18 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
+        execution_context ? execution_context->GetTargetSP() : TargetSP();
+    BreakpointSP bp_sp = target_sp->CreateBreakpoint(
+        nullptr, nullptr, "main", eFunctionNameTypeAuto, eLanguageTypeUnknown,
+        0, eLazyBoolNo, false, false);
```
Don't you want eFunctionNameTypeFull?  You wouldn't want to break on SomeCPlusPlusClass::main, which eFunctionNameTypeBase will break on.

clayborg replied:
Yes, you are correct, eFunctionNameTypeFull is the right choice. 

Jim what do you think about getting the platform involved in setting this breakpoint as I described? Is that overkill? I am good with either approach.

jimingham replied:
I think abstracting out each Language's entry point function name is the only really strong argument for adding this option, so I think we should do that.  But as I said below, I think that's something our supported Language plugins would know, I don't think the Platform knows this.

jimingham commented on lldb/source/Commands/CommandOptionsProcessLaunch.cpp:
```diff
@@ -38,7 +38,18 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
+    TargetSP target_sp =
+        execution_context ? execution_context->GetTargetSP() : TargetSP();
+    BreakpointSP bp_sp = target_sp->CreateBreakpoint(
+        nullptr, nullptr, "main", eFunctionNameTypeAuto, eLanguageTypeUnknown,
```
IIUC, we're not trying to break at the actual entry point (that's what -stop-at-entry-point does) but the main executable's Language entry point.  I'm not sure that is something the platform can know.   For all the C-based languages it was "main", but for some other languages it might be something else.  For Ada apparently it can be anything and you designate it "main" some other way.  In other languages it could just have a different name than "main"...

If that's true, shouldn't we be asking all the Languages we support what their "main" function is, not the Platforms.  I'm not sure they can know.  And for situations like the Ada one the language would have to be able to figure out what the main function was by some Ada specific means.

jimingham replied:
In the simplest case, we gather a set of names one or more from each Language, and then make one multiple name breakpoint for all of them.

clayborg replied:
Yeah, there is no way to know what language the main entrypoint was written in, so no hope of figuring out the language. I would be fine with the multiple names approach from all languages. If you want that feature, can you give Jose some clues what he should do to make that happen?

jimingham replied:
We have a list of "Supported Languages" (Language::GetSupportedLangauges) which are the ones we actually have plugins for as opposed to the ones whose names we know.  That's the list you should be querying. 

We should add a new API to the Language plugin: Language::GetUserEntrypointName() and then this code should iterate over all the supported languages building up a unique list of entry point names using this API.  At present there's going to be one: "main" repeated many times but I think Walter promised us another one soon...  

If you find more than one name, there's a Target::CreateBreakpoint overload that takes a `const char *func_names[]` - that you can use to make one portmanteau breakpoint for all the entry point names.  

It's also a good idea to add a shared library filter instead of the first `nullptr` in the code above, and specify the main executable.  You know the entry point has to be in the main executable, and so it would be inefficient to search the whole world for this breakpoint.  It could also end up not being what you want, because some other shared library might call a function named "main" as part of it's shared library loading code (which happens before the language's entry point) so you would in fact stop too early.

You'll need to pass eLanguageTypeUnknown for the language.

jimingham replied:
It's probably also fine to implement a default version in the Language class that returns `main`.  No need to make everyone override this method when it's `main` for most languages.

junior-jl replied:
> We should add a new API to the Language plugin: Language::GetUserEntrypointName() and then this code should iterate over all the supported languages building up a unique list of entry point names using this API. At present there's going to be one: "main" repeated many times but I think Walter promised us another one soon...

About this, Walter had proposed the following: "I'd suggest adding a virtual method GetMainSymbol to the Language.h interface, which could be implemented by each language plugin.". I guess it's a good approach. What are your thoughts?

junior-jl replied:
> It's also a good idea to add a shared library filter instead of the first `nullptr` in the code above, and specify the main executable. You know the entry point has to be in the main executable, and so it would be inefficient to search the whole world for this breakpoint. It could also end up not being what you want, because some other shared library might call a function named "main" as part of it's shared library loading code (which happens before the language's entry point) so you would in fact stop too early.

I see. Something like this?

```cpp
ModuleSP main_module_sp = target_sp->GetExecutableModule();
FileSpecList shared_lib_filter;
shared_lib_filter.Append(main_module_sp->GetFileSpec());
BreakpointSP bp_sp = target_sp->CreateBreakpoint(
    &shared_lib_filter, nullptr, "main", eFunctionNameTypeFull, eLanguageTypeUnknown,
    0, eLazyBoolNo, false, false);
```

jimingham commented on lldb/source/Commands/Options.td:
```diff
@@ -675,6 +675,8 @@ let Command = "platform shell" in {
 let Command = "process launch" in {
   def process_launch_stop_at_entry : Option<"stop-at-entry", "s">,
     Desc<"Stop at the entry point of the program when launching a process.">;
+  def process_launch_stop_at_main : Option<"stop-at-main", "m">,
+    Desc<"Stop at the main function of the program when launching a process.">;
```
Yes, that better expresses what this does.

junior-jl replied:
Okay. I'll change this. Should I keep 'm' as the short option?

clayborg reviewed (commented):
After Jim's recent comments sounds like the Platform is not the right place to set the breakpoint. So please ignore my comments suggesting to use the Platform layer.

jimingham commented:
Yes that's right.  Normally what happens in program execution is the main executable gets loaded along with the dynamic loader, and then the loader runs, leading the program through several rounds of shared library loading (that can be a lot of shared libraries these days) and THEN runs the user entry point.  Adding this filter means that we won't try to resolve a "main" breakpoint in every one of those shared libraries.

Jim

> On Sep 26, 2023, at 7:39 AM, José Lira Junior ***@***.***> wrote:
> 
> 
> @junior-jl commented on this pull request.
> 
> In lldb/source/Commands/CommandOptionsProcessLaunch.cpp <https://github.com/llvm/llvm-project/pull/67019#discussion_r1337326182>:
> 
> > @@ -38,7 +38,18 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
>    case 's': // Stop at program entry point
>      launch_info.GetFlags().Set(eLaunchFlagStopAtEntry);
>      break;
> -
> +  case 'm': // Stop at main function
> +  {
> +    TargetSP target_sp =
> +        execution_context ? execution_context->GetTargetSP() : TargetSP();
> +    BreakpointSP bp_sp = target_sp->CreateBreakpoint(
> +        nullptr, nullptr, "main", eFunctionNameTypeAuto, eLanguageTypeUnknown,
> It's also a good idea to add a shared library filter instead of the first nullptr in the code above, and specify the main executable. You know the entry point has to be in the main executable, and so it would be inefficient to search the whole world for this breakpoint. It could also end up not being what you want, because some other shared library might call a function named "main" as part of it's shared library loading code (which happens before the language's entry point) so you would in fact stop too early.
> 
> I see. Something like this?
> 
> ModuleSP main_module_sp = target_sp->GetExecutableModule();
> FileSpecList shared_lib_filter;
> shared_lib_filter.Append(main_module_sp->GetFileSpec());
> BreakpointSP bp_sp = target_sp->CreateBreakpoint(
>     &shared_lib_filter, nullptr, "main", eFunctionNameTypeFull, eLanguageTypeUnknown,
>     0, eLazyBoolNo, false, false);
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/67019#discussion_r1337326182>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW5NFBYC7PEIMKPRGZTX4LSIZANCNFSM6AAAAAA5BUJK5M>.
> You are receiving this because you were mentioned.
>

jimingham commented:
> On Sep 26, 2023, at 7:29 AM, José Lira Junior ***@***.***> wrote:
> 
> 
> @junior-jl commented on this pull request.
> 
> In lldb/source/Commands/CommandOptionsProcessLaunch.cpp <https://github.com/llvm/llvm-project/pull/67019#discussion_r1337311231>:
> 
> > @@ -38,7 +38,18 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
>    case 's': // Stop at program entry point
>      launch_info.GetFlags().Set(eLaunchFlagStopAtEntry);
>      break;
> -
> +  case 'm': // Stop at main function
> +  {
> +    TargetSP target_sp =
> +        execution_context ? execution_context->GetTargetSP() : TargetSP();
> +    BreakpointSP bp_sp = target_sp->CreateBreakpoint(
> +        nullptr, nullptr, "main", eFunctionNameTypeAuto, eLanguageTypeUnknown,
> We should add a new API to the Language plugin: Language::GetUserEntrypointName() and then this code should iterate over all the supported languages building up a unique list of entry point names using this API. At present there's going to be one: "main" repeated many times but I think Walter promised us another one soon...
> 
> About this, Walter had proposed the following: "I'd suggest adding a virtual method GetMainSymbol to the Language.h interface, which could be implemented by each language plugin.". I guess it's a good approach. What are your thoughts?
> 

Right, that's also what I was suggesting.  W.r.t. Walter's suggestion, I still think it's better to use a generic term (I suggested GetUserEntrypointName) rather than explicitly using  the word "main", which is the C name for this feature...  Greg suggested making this change to the option name as well earlier in this thread.

Jim

> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/67019#discussion_r1337311231>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW42YJDE57AFPW3Y7CDX4LRFPANCNFSM6AAAAAA5BUJK5M>.
> You are receiving this because you were mentioned.
>

junior-jl commented:
This is the case in the current commit. I was trying to implement this with SmallSet and const char* to use the overload you mentioned, Jim, but I was having trouble with it, specially because I needed to hard code a 'max number of languages'. So I used the overload that takes a std::vector<std::string>. Let me know if that's not good. 

```cpp
case 'm': // Stop at user entry point
  {
    TargetSP target_sp =
        execution_context ? execution_context->GetTargetSP() : TargetSP();
    ModuleSP main_module_sp = target_sp->GetExecutableModule();
    FileSpecList shared_lib_filter;
    shared_lib_filter.Append(main_module_sp->GetFileSpec());
    std::vector<std::string> entryPointNames;
    for (LanguageType lang_type : Language::GetSupportedLanguages()) {
      Language *lang = Language::FindPlugin(lang_type);
      if (lang) {
        std::string entryPointName = lang->GetUserEntryPointName();
        if (!entryPointName.empty()) {
          entryPointNames.push_back(entryPointName);
        }
      }
    }
    BreakpointSP bp_sp = target_sp->CreateBreakpoint(
        /*containingModules=*/&shared_lib_filter,
        /*containingSourceFiles=*/nullptr, 
        /*func_names=*/entryPointNames,
        /*func_name_type_mask=*/eFunctionNameTypeFull,
        /*language=*/eLanguageTypeUnknown,
        /*offset=*/0, /*skip_prologue=*/eLazyBoolNo, /*internal=*/false,
        /*hardware=*/false);
    if (!bp_sp)
      error.SetErrorString("Breakpoint creation failed.\n");
    bp_sp->SetOneShot(true);
    break;
  }
```

junior-jl commented:
Also, I just noticed that `clang-format` did some minor changes to `Language.h`. Please, let me know if I need to revert those.

clayborg commented on lldb/source/Commands/CommandOptionsProcessLaunch.cpp:
```diff
@@ -38,7 +39,36 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
+      Language *lang = Language::FindPlugin(lang_type);
+      if (lang) {
+        std::string entryPointName = lang->GetUserEntryPointName();
+        if (!entryPointName.empty()) {
+          entryPointNames.push_back(entryPointName);
+        }
```
No parens on single line if statements per llvm coding guidelines

clayborg commented on lldb/source/Commands/CommandOptionsProcessLaunch.cpp:
```diff
@@ -38,7 +39,36 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
+    ModuleSP main_module_sp = target_sp->GetExecutableModule();
+    FileSpecList shared_lib_filter;
+    shared_lib_filter.Append(main_module_sp->GetFileSpec());
+    std::vector<std::string> entryPointNames;
```
Do we end up with many duplicate entries in the std::vector after iterating over all languages? If so you can use a `std::set<std::string>` so we can dedupe the names, then before you set the breakpoint convert the `std::set<std::string>` into a `std::vector<string>` with:
```
std::vector<std::string> entryPointNames(entryPointNameSet.begin(), entryPointNameSet.end());
```

clayborg commented on lldb/source/Commands/Options.td:
```diff
@@ -675,6 +675,8 @@ let Command = "platform shell" in {
   def process_launch_stop_at_entry : Option<"stop-at-entry", "s">,
     Desc<"Stop at the entry point of the program when launching a process.">;
+  def process_launch_stop_at_main : Option<"stop-at-user-entry", "m">,
+    Desc<"Stop at the user entry point when launching a process.">;
```
It would be fine to mention "main" here something like:
```
Desc<"Stop at the user entry point when launching a process. For C based languages this will be the 'main' function, but this might differ for other languages.">;
```

jimingham commented:
We try to only add really frequently used aliases to the default lldb set.  In particular, the "unique first one and two character" space in the lldb command set is valuable, that's for your most common every-day commands, and we try to leave as much of it open as possible for users to customize lldb for their own purposes.

In the case of "start", that overlaps with "step" so it doesn't actually take up a new slot in that one or two character space, so it's not so bad.  However, we certainly don't want to make any of the partial spellings of "step" become ambiguous.  In the lldb command line exact matches to aliases always win.  There's one for `s` but it doesn't look like there's one for `st`.  So if you are going to add `start` as an alias, you should also add an "st" alias to "thread step" since that's a much more common command.

I personally wouldn't use "start" enough to warrant having it in the default command set.  

One of the main reasons that "stopping at main" was handy in gdb was that gdb didn't originally do "future-break" so you always had to run to main to get everything loaded, then set your breakpoints.  That's not necessary in lldb (or really modern gdb's either)...  I think if you are using toy programs, stopping at main is pretty common, but I don't think it's all that common once you are debugging significant programs, main is generally pretty far from anything interesting.

But I don't have a strong opinion about that, if other people want to weigh in.

And to answer your direct question, you are right, you add aliases to the default command set by putting a new AddAlias in CommandInterpreter.cpp.

Jim


> On Sep 27, 2023, at 7:31 AM, José Lira Junior ***@***.***> wrote:
> 
> 
> What are your thoughts on creating an alias for process launch --stop-at-user-entry such as the intended start (or begin, or even run -m).
> 
> Also, if I wanted to do that, the only change would be an AddAlias in CommandInterpreter.cpp? Or there's something else?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/67019#issuecomment-1737522305>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW7K226YIHINLF4CR7TX4Q2FLANCNFSM6AAAAAA5BUJK5M>.
> You are receiving this because you were mentioned.
>

junior-jl commented:
I understand, thank you, Jim.

Is there anything else I should change in this PR?

jimingham commented on lldb/source/Commands/CommandOptionsProcessLaunch.cpp:
```diff
@@ -38,7 +39,37 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
+    std::vector<std::string> entryPointNames(entryPointNamesSet.begin(),
+                                             entryPointNamesSet.end());
+    BreakpointSP bp_sp = target_sp->CreateBreakpoint(
+        /*containingModules=*/&shared_lib_filter,
```
We tend to put this sort of doc comment after the variable rather than before, and don't put it on arguments where the variable name being passed (e.g. `shared_lib_filter`) is self-documenting.

junior-jl replied:
You mean like this?

```cpp
BreakpointSP bp_sp = target_sp->CreateBreakpoint(
    &shared_lib_filter, 
    nullptr,                 // containingSourceFiles
    entryPointNames,     
    eFunctionNameTypeFull,   // func_name_type_mask
    eLanguageTypeUnknown,    // language
    0,                       // offset
    eLazyBoolNo,             // skip_prologue
    false                    // internal
    false                    // hardware
);
```

jimingham replied:
Yes.

jimingham reviewed (approved):
Other than that trivial doc style comment, this LGTM.

junior-jl commented:
In this last commit, I changed the comments style and used @Michael137 suggestion by declaring a `llvm::SetVector` instead of a temporary `std::set`.

junior-jl commented on lldb/source/Commands/CommandOptionsProcessLaunch.cpp:
```diff
@@ -38,7 +39,37 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
+          entryPointNamesSet.insert(entryPointName);
+      }
+    }
+    std::vector<std::string> entryPointNames(entryPointNamesSet.begin(),
```
Done. Thank you for the help.

medismailben commented on lldb/source/Commands/CommandOptionsProcessLaunch.cpp:
```diff
@@ -38,7 +40,40 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
+        entryPointNamesSet;
+    for (LanguageType lang_type : Language::GetSupportedLanguages()) {
+      Language *lang = Language::FindPlugin(lang_type);
+      if (lang) {
```
You could inverse the condition and return early.

medismailben commented on lldb/source/Commands/CommandOptionsProcessLaunch.cpp:
```diff
@@ -38,7 +40,40 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
+        if (!entryPointName.empty())
+          entryPointNamesSet.insert(entryPointName);
+      }
+    }
```
You could check the size of `entryPointNamesSet` before trying to create the breakpoint, and return early if it's empty.
Breakpoint creation could fail for many reasons, so if you couldn't get any entry point name from the language plugin, we shouldn't bother creating a breakpoint and give the user a more accurate error message.

medismailben commented on lldb/include/lldb/Target/Language.h:
```diff
@@ -160,6 +166,10 @@ class Language : public PluginInterface {
 
+  // Implement this function to return the user-defined entry point name
+  // for the language
+  virtual const char *GetUserEntryPointName() const { return nullptr; }
```
nit: this could have returned an `llvm::StringRef` instead of an `const char *`.

medismailben replied:
```suggestion
  virtual llvm::StringRef GetUserEntryPointName() const { return {}; }
```

medismailben commented on lldb/source/Plugins/Language/CPlusPlus/CPlusPlusLanguage.h:
```diff
@@ -103,6 +103,8 @@ class CPlusPlusLanguage : public Language {
     return lldb::eLanguageTypeC_plus_plus;
   }
 
+  const char *GetUserEntryPointName() const override { return "main"; }
```
same comment as earlier regarding the return type.

medismailben commented on lldb/source/Plugins/Language/ObjC/ObjCLanguage.h:
```diff
@@ -127,6 +127,8 @@ class ObjCLanguage : public Language {
     return lldb::eLanguageTypeObjC;
   }
 
+  const char *GetUserEntryPointName() const override { return "main"; }
```
ditto.

medismailben commented on lldb/source/Plugins/Language/ObjCPlusPlus/ObjCPlusPlusLanguage.h:
```diff
@@ -27,6 +27,8 @@ class ObjCPlusPlusLanguage : public Language {
     return lldb::eLanguageTypeObjC_plus_plus;
   }
 
+  const char *GetUserEntryPointName() const override { return "main"; }
```
Unrelated to this PR but it would have been nice if this class was derived from the `CPlusPlus` and/or `ObjC` one so you wouldn't have to re-implement the same method here. @Michael137 do you think that it could be a useful refactor ? Or may be we should introduce a `C` language base class that every thing derives from ?

medismailben reviewed (approved):
LGTM with some comments.

junior-jl commented:
Hi, @medismailben. I changed the code with your suggestions. Can you take a look if the early exits are correct before I commit, please?

```cpp
case 'm': // Stop at user entry point
  {
    TargetSP target_sp =
        execution_context ? execution_context->GetTargetSP() : TargetSP();
    ModuleSP main_module_sp = target_sp->GetExecutableModule();
    FileSpecList shared_lib_filter;
    shared_lib_filter.Append(main_module_sp->GetFileSpec());
    llvm::SetVector<std::string, std::vector<std::string>, std::unordered_set<std::string>> entryPointNamesSet;
    for (LanguageType lang_type : Language::GetSupportedLanguages()) {
      Language *lang = Language::FindPlugin(lang_type);
      if (!lang) {
        error.SetErrorString("Language not found\n");
        break;
      }
      std::string entryPointName = lang->GetUserEntryPointName().str();
      if (!entryPointName.empty())
        entryPointNamesSet.insert(entryPointName);
    }
    if (entryPointNamesSet.empty()) {
      error.SetErrorString("No entry point name found\n");
      break;
    }
    BreakpointSP bp_sp = target_sp->CreateBreakpoint(
    &shared_lib_filter, 
    nullptr,                 // containingSourceFiles
    entryPointNamesSet.takeVector(),     
    eFunctionNameTypeFull,   // func_name_type_mask
    eLanguageTypeUnknown,    // language
    0,                       // offset
    eLazyBoolNo,             // skip_prologue
    false,                   // internal
    false                    // hardware
    );
    if (!bp_sp) {
      error.SetErrorString("Breakpoint creation failed.\n");
      break;
    }
    bp_sp->SetOneShot(true);
    break;
  }
```

Also, thank you for the reminder of changing the description of the PR. Done :)

junior-jl commented:
Oh, no, I was following [this](https://llvm.org/docs/GitHub.html#landing-your-change), but it looks like I messed up, can someone help me to correct this, please?
```
Before merging the PR, it is recommended that you rebase locally and re-run test checks:

# Add upstream as a remote (if you don't have it already)
git remote add upstream https://github.com/llvm/llvm-project.git

# Make sure you have all the latest changes
git fetch upstream && git rebase -i upstream/main

# Make sure tests pass with latest changes and your change
ninja check

# Push the rebased changes to your fork.
git push origin my_change -f
```

clayborg commented:
> Hi, @medismailben. I changed the code with your suggestions. Can you take a look if the early exits are correct before I commit, please?
> 
> ```c++
> case 'm': // Stop at user entry point
>   {
>     TargetSP target_sp =
>         execution_context ? execution_context->GetTargetSP() : TargetSP();
>     ModuleSP main_module_sp = target_sp->GetExecutableModule();
>     FileSpecList shared_lib_filter;
>     shared_lib_filter.Append(main_module_sp->GetFileSpec());
>     llvm::SetVector<std::string, std::vector<std::string>, std::unordered_set<std::string>> entryPointNamesSet;
>     for (LanguageType lang_type : Language::GetSupportedLanguages()) {
>       Language *lang = Language::FindPlugin(lang_type);
>       if (!lang) {
>         error.SetErrorString("Language not found\n");
>         break;
>       }
>       std::string entryPointName = lang->GetUserEntryPointName().str();
>       if (!entryPointName.empty())
>         entryPointNamesSet.insert(entryPointName);
>     }
>     if (entryPointNamesSet.empty()) {
>       error.SetErrorString("No entry point name found\n");
>       break;
>     }
>     BreakpointSP bp_sp = target_sp->CreateBreakpoint(
>     &shared_lib_filter, 
>     nullptr,                 // containingSourceFiles
>     entryPointNamesSet.takeVector(),     
>     eFunctionNameTypeFull,   // func_name_type_mask
>     eLanguageTypeUnknown,    // language
>     0,                       // offset
>     eLazyBoolNo,             // skip_prologue
>     false,                   // internal
>     false                    // hardware
>     );
>     if (!bp_sp) {
>       error.SetErrorString("Breakpoint creation failed.\n");
>       break;
>     }
>     bp_sp->SetOneShot(true);
>     break;
>   }
> ```
> 
> Also, thank you for the reminder of changing the description of the PR. Done :)

This code in this switch statement should go into Target.h/Target.cpp as something like:
```
lldb::BreakpointSP Target::CreateBreakpointAtUserEntry();
```
Then the code in this switch just calls that function. Then we can expose this through the SBTarget API as well.

medismailben reviewed (changes requested):
Please fix your branch before merging this PR.

junior-jl commented:
> @junior-jl You pushed 921 commits to your branch, I doubt you wanted to do this.
> 
> To fix this, I'd first find all your commits (so you don't lose your previous work):
> 
> ```
> $ git log --author "José Lira Junior" # or your email address
> ```
> 
> Save the commit hashes to recover them later on.
> 
> I'd reset your branch to main:
> 
> ```
> $ git checkout start
> $ git fetch origin
> $ git reset --hard origin/main
> ```
> 
> Then, you just have to cherry-pick the commits you saved earlier:
> 
> ```
> $ git cherry-pick <commit-hash>
> ```
> 
> Finally, push your branch again to your personal GitHub remote:
> 
> ```
> $ git push -f <personal-github-remote> start
> ```
> 
> I think you can also squash all the commits into a single one to reduce the noise and make it easier to revert in case your patch causes some issues when it lands.

Oh, thank you so much!

I guess it is correct now.

junior-jl commented:
> > Hi, @medismailben. I changed the code with your suggestions. Can you take a look if the early exits are correct before I commit, please?
> > ```c++
> > case 'm': // Stop at user entry point
> >   {
> >     TargetSP target_sp =
> >         execution_context ? execution_context->GetTargetSP() : TargetSP();
> >     ModuleSP main_module_sp = target_sp->GetExecutableModule();
> >     FileSpecList shared_lib_filter;
> >     shared_lib_filter.Append(main_module_sp->GetFileSpec());
> >     llvm::SetVector<std::string, std::vector<std::string>, std::unordered_set<std::string>> entryPointNamesSet;
> >     for (LanguageType lang_type : Language::GetSupportedLanguages()) {
> >       Language *lang = Language::FindPlugin(lang_type);
> >       if (!lang) {
> >         error.SetErrorString("Language not found\n");
> >         break;
> >       }
> >       std::string entryPointName = lang->GetUserEntryPointName().str();
> >       if (!entryPointName.empty())
> >         entryPointNamesSet.insert(entryPointName);
> >     }
> >     if (entryPointNamesSet.empty()) {
> >       error.SetErrorString("No entry point name found\n");
> >       break;
> >     }
> >     BreakpointSP bp_sp = target_sp->CreateBreakpoint(
> >     &shared_lib_filter, 
> >     nullptr,                 // containingSourceFiles
> >     entryPointNamesSet.takeVector(),     
> >     eFunctionNameTypeFull,   // func_name_type_mask
> >     eLanguageTypeUnknown,    // language
> >     0,                       // offset
> >     eLazyBoolNo,             // skip_prologue
> >     false,                   // internal
> >     false                    // hardware
> >     );
> >     if (!bp_sp) {
> >       error.SetErrorString("Breakpoint creation failed.\n");
> >       break;
> >     }
> >     bp_sp->SetOneShot(true);
> >     break;
> >   }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Also, thank you for the reminder of changing the description of the PR. Done :)
> 
> This code in this switch statement should go into Target.h/Target.cpp as something like:
> 
> ```
> lldb::BreakpointSP Target::CreateBreakpointAtUserEntry();
> ```
> 
> Then the code in this switch just calls that function. Then we can expose this through the SBTarget API as well.

Done! ✅

medismailben commented on lldb/source/Target/Target.cpp:
```diff
@@ -782,7 +824,7 @@ void Target::ApplyNameToBreakpoints(BreakpointName &bp_name) {
 void Target::GetBreakpointNames(std::vector<std::string> &names) {
   names.clear();
-  for (const auto& bp_name_entry : m_breakpoint_names) {
+  for (const auto &bp_name_entry : m_breakpoint_names) {
```
All the formatting below seems unrelated to your PR, probably because you ran `clang-format` on the whole file. We try to avoid this because is overwrites the file history with the actual explanations of what previous contributors changes, with your commit which is not related.

If you want to reformat the rest of the file, that should be done in a separate commit and you should add the commit hash to the [.git-blame-ignore-revs](https://github.com/llvm/llvm-project/blob/main/.git-blame-ignore-revs) file.

Now, if you want to address this, you can do the following:

0. Save your current branch in case things go wrong with git:
```
$ git checkout -b stop-at-user-entry-formatted
```

1. Un-do commit without deleting your work.
```
$ git reset HEAD~1
```
2. Add only changes related to the PR (skip all the formatting changes).
```
$ git add -p .
```
3. Make sure the changes that are staged don't contain reformatting changes.
```
$ git status
$ git commit -v
```
4. Discard the formatting changes.
```
$ git reset --hard HEAD
```

If you want to format specifically what you changed in your commit, you should run:
```
$ git clang-format HEAD~
$ git add -u
$ git commit --amend --no-edit
```
Make sure you don't have any other unstaged changes before doing this, otherwise, they'll be merged with the formatting changes.

Let me know if you need help or have any questions.

junior-jl replied:
I was wondering if that was a problem, thank you for clarifying and for the step-by-step!

Should I squash the two commits again?

medismailben replied:
> Should I squash the two commits again?

You can if you want.

medismailben commented:
I think you're good for the implementing however it would be nice to add a shell test to make sure it works well and it doesn't break in the future.

medismailben commented:
You can use `lldb/test/Shell/Commands/command-thread-select.test` as an example to write your own test.

junior-jl commented:
> You can use `lldb/test/Shell/Commands/command-thread-select.test` as an example to write your own test.

Using this file as a template, I wrote `command-process-launch-user-entry.test` as shown below:

```
# RUN: %clang_host -g %S/Inputs/main.c -o %t
# RUN: %lldb %t -s %s -o exit | FileCheck %s

process launch -m
# CHECK-LABEL: process launch -m
# CHECK: Process {{.*}} stopped
# CHECK: stop reason = one-shot breakpoint 1
# CHECK:   frame #0: {{.*}}`main at main.c
```

But I wanted to add the check line `# CHECK: Process {{.*}} launched: '{{.*}}' {{.*}}`, but this fails because the following is happening:

1. When the command is executed from the `(lldb)` prompt, it correctly shows that the process is launched then stopped.

```
$ ./bin/lldb ~/main.out
(lldb) target create "/home/jose/main.out"
Current executable set to '/home/jose/main.out' (x86_64).
(lldb) process launch -m
Process 63632 launched: '/home/jose/main.out' (x86_64)
Process 63632 stopped
* thread #1, name = 'main.out', stop reason = one-shot breakpoint 1
    frame #0: 0x0000555555555140 main.out`main at main.c:2
   1   	int foo() { return 0; }
-> 2   	int main() { return foo(); }
   3   	
(lldb) 
```

2. When I run like shown below, the `Process launched` message is shown after the stop.

```
$ ./bin/lldb ~/main.out -o 'process launch -m'
(lldb) target create "/home/jose/main.out"
Current executable set to '/home/jose/main.out' (x86_64).
(lldb) process launch -m
Process 63846 stopped
* thread #1, name = 'main.out', stop reason = one-shot breakpoint 1
    frame #0: 0x0000555555555140 main.out`main at main.c:2
   1   	int foo() { return 0; }
-> 2   	int main() { return foo(); }
   3
Process 63846 launched: '/home/jose/main.out' (x86_64)
(lldb) 
```

Is this behaviour expected?

medismailben commented:
> > You can use `lldb/test/Shell/Commands/command-thread-select.test` as an example to write your own test.
> 
> Using this file as a template, I wrote `command-process-launch-user-entry.test` as shown below:
> 
> ```
> # RUN: %clang_host -g %S/Inputs/main.c -o %t
> # RUN: %lldb %t -s %s -o exit | FileCheck %s
> 
> process launch -m
> # CHECK-LABEL: process launch -m
> # CHECK: Process {{.*}} stopped
> # CHECK: stop reason = one-shot breakpoint 1
> # CHECK:   frame #0: {{.*}}`main at main.c
> ```
> 
> But I wanted to add the check line `# CHECK: Process {{.*}} launched: '{{.*}}' {{.*}}`, but this fails because the following is happening:
> 
> 1. When the command is executed from the `(lldb)` prompt, it correctly shows that the process is launched then stopped.
> 
> ```
> $ ./bin/lldb ~/main.out
> (lldb) target create "/home/jose/main.out"
> Current executable set to '/home/jose/main.out' (x86_64).
> (lldb) process launch -m
> Process 63632 launched: '/home/jose/main.out' (x86_64)
> Process 63632 stopped
> * thread #1, name = 'main.out', stop reason = one-shot breakpoint 1
>     frame #0: 0x0000555555555140 main.out`main at main.c:2
>    1   	int foo() { return 0; }
> -> 2   	int main() { return foo(); }
>    3   	
> (lldb) 
> ```
> 
> 2. When I run like shown below, the `Process launched` message is shown after the stop.
> 
> ```
> $ ./bin/lldb ~/main.out -o 'process launch -m'
> (lldb) target create "/home/jose/main.out"
> Current executable set to '/home/jose/main.out' (x86_64).
> (lldb) process launch -m
> Process 63846 stopped
> * thread #1, name = 'main.out', stop reason = one-shot breakpoint 1
>     frame #0: 0x0000555555555140 main.out`main at main.c:2
>    1   	int foo() { return 0; }
> -> 2   	int main() { return foo(); }
>    3
> Process 63846 launched: '/home/jose/main.out' (x86_64)
> (lldb) 
> ```
> 
> Is this behaviour expected?

I can reproduce it:

```
$ ./bin/lldb /tmp/main -o "b main" -o "r"
(lldb) target create "/tmp/main"
Current executable set to '/tmp/main' (arm64).
(lldb) b main
Breakpoint 1: where = main`main + 12 at main.cpp:2:3, address = 0x0000000100003fa0
(lldb) r
Process 53764 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x0000000100003fa0 main`main at main.cpp:2:3
   1   	int main() {
-> 2   	  return 0;
   3   	}
Process 53764 launched: '/tmp/main' (arm64)
```

By default the debugger runs in asynchronous mode so the stop events can be handled in a nondetermistic way. However, I think this confusing and we should do something about (in a separate PR). Could you file a new issue describing this behavior and pasting the link here.

Your test looks fine to me, so we can move on with it to merge your PR. I'll update it after we fix that other bug.

junior-jl commented:
> Could you file a new issue describing this behavior and pasting the link here.

Done! https://github.com/llvm/llvm-project/issues/68035

> Your test looks fine to me, so we can move on with it to merge your PR. I'll update it after we fix that other bug.

Great. Added the test in the latest commit.

medismailben commented:
@clayborg what do you think ? Can we merge this ?

JDevlieghere commented on lldb/include/lldb/Target/Language.h:
```diff
@@ -95,21 +95,24 @@ class Language : public PluginInterface {
   public:
     EitherTypeScavenger() : TypeScavenger() {
-      for (std::shared_ptr<TypeScavenger> scavenger : { std::shared_ptr<TypeScavenger>(new ScavengerTypes())... }) {
+      for (std::shared_ptr<TypeScavenger> scavenger :
+           {std::shared_ptr<TypeScavenger>(new ScavengerTypes())...}) {
```
Can this use `make_shared`?

medismailben replied:
FWIW, these changes were not introduced by @junior-jl, they're probably the result of running `clang-format` on the whole file.

junior-jl replied:
Yes, that was indeed the case. I believe I removed all the unrelated modifications from the PR in this last push.

JDevlieghere commented on lldb/source/Target/Target.cpp:
```diff
@@ -335,6 +335,48 @@ BreakpointSP Target::GetBreakpointByID(break_id_t break_id) {
+  for (LanguageType lang_type : Language::GetSupportedLanguages()) {
+    Language *lang = Language::FindPlugin(lang_type);
+    if (!lang) {
+      error.SetErrorString("Language not found\n");
```
The `error` is a local variable which nobody is reading. Should this be an out parameter?

junior-jl replied:
Oh, that's true. If I understood correctly, the method should be `lldb::BreakpointSP lldb_private::Target::CreateBreakpointAtUserEntry(Status &error)` and in `CommandOptionsProcessLaunch.cpp`, it should be called with `target_sp->CreateBreakpointAtUserEntry(error);`. Am I correct?

kastiglione replied:
that's correct.

junior-jl replied:
Done ✅

JDevlieghere commented on lldb/source/Commands/CommandOptionsProcessLaunch.cpp:
```diff
@@ -38,7 +41,13 @@ Status CommandOptionsProcessLaunch::SetOptionValue(
-
+  case 'm': // Stop at user entry point
+  {
+    TargetSP target_sp =
+        execution_context ? execution_context->GetTargetSP() : TargetSP();
```
There's another option below (`'a'`) that also needs the `target_sp`. Let's hoist this out of the switch and reuse it in all the options.

junior-jl committed:
[lldb] add stop-at-user-entry option to process launch

[lldb] add stop-at-user-entry option to process launch

refactor CreateBreakpointAtUserEntry method

remove unecessary includes

kastiglione commented on lldb/source/Target/Target.cpp:
```diff
@@ -335,6 +335,48 @@ BreakpointSP Target::GetBreakpointByID(break_id_t break_id) {
 }
 
+lldb::BreakpointSP lldb_private::Target::CreateBreakpointAtUserEntry() {
+  TargetSP target_sp = shared_from_this();
```
Why is this necessary?

junior-jl replied:
I removed that line and the use of `target_sp`, hence, these lines also changed:

```cpp
ModuleSP main_module_sp = GetExecutableModule();
BreakpointSP bp_sp = CreateBreakpoint(...)
```

It still works normally and pass the test. Thank you.

clayborg commented on lldb/source/Commands/CommandOptionsProcessLaunch.cpp:
```diff
@@ -8,18 +8,21 @@
 #include "lldb/Target/Target.h"
-
 #include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SetVector.h"
```
we can remove this and the above includes now that we are doing the heavy work in Target.cpp

junior-jl replied:
Done!

clayborg reviewed (commented):
Just remove the includes that are no longer needed and this is good to go!

medismailben commented:
@junior-jl I think everyone approved the PR so you just need to address the formatting issues so we can merge it. Would you mind squashing your 3 commits into a single one and running `git clang-format HEAD~` and `git commit --amend --no-edit` and pushing your branch again ? Thanks!

JDevlieghere commented on lldb/include/lldb/Target/Language.h:
```diff
@@ -160,6 +160,10 @@ class Language : public PluginInterface {
 
   virtual lldb::LanguageType GetLanguageType() const = 0;
 
+  // Implement this function to return the user-defined entry point name
+  // for the language
```
Missing period

JDevlieghere commented on lldb/source/Target/Target.cpp:
```diff
@@ -335,6 +335,47 @@ BreakpointSP Target::GetBreakpointByID(break_id_t break_id) {
+    return lldb::BreakpointSP();
+  }
+  BreakpointSP bp_sp =
+      CreateBreakpoint(&shared_lib_filter,
+                       nullptr, // containingSourceFiles
+                       entryPointNamesSet.takeVector(),
+                       eFunctionNameTypeFull, // func_name_type_mask
+                       eLanguageTypeUnknown,  // language
+                       0,                     // offset
+                       eLazyBoolNo,           // skip_prologue
+                       false,                 // internal
+                       false                  // hardware
```
Is there prior art in Target.cpp for specifying the arguments like this? I think most of the code base uses inline comments (e.g. `/*containingSourceFiles=*/nullptr`)

JDevlieghere commented on lldb/test/Shell/Commands/command-process-launch-user-entry.test:
```diff
@@ -0,0 +1,8 @@
+# CHECK-LABEL: process launch -m
+# CHECK: Process {{.*}} stopped
+# CHECK: stop reason = one-shot breakpoint 1
+# CHECK:   frame #0: {{.*}}`main at main.c
```
No newline at the end of this file?

junior-jl commented:
Addressed @JDevlieghere comments regarding formatting and inlined comments.

JDevlieghere reviewed (approved):
LGTM!

clayborg commented:
very nice! Thanks for doing all of our suggested improvements, great patch. I would love to see a new API in SBTarget that exposes this feature:
```
lldb::SBBreakpoint lldb::SBTarget::CreateBreakpointAtUserEntry();
```
